docs: Docs for sensitive variables (#26431)
hashicorp/terraform: website/docs/configuration/variables.html.md

value from within the module.

The name of a variable can be any valid [identifier](./syntax.html#identifiers)

> _except_ the following: `source`, `version`, `providers`, `count`, `for_each`, `lifecycle`, `depends_on`, `locals`.

Reviewer: I'm curious about this change;  these are in-page links on the website (I believe that's what the extra `-` does): 

https://www.terraform.io/docs/configuration/variables.html#declaring-an-input-variable

Author: It is solely to make them take up less real-estate on the page. They are links, but you'll note that they just link to themselves ... which adds no extra context. https://www.terraform.io/docs/configuration/variables.html#locals

Reviewer: got it, thanks!

These names are reserved for meta-arguments in
[module configuration blocks](./modules.html), and cannot be
declared as variable names.

## Arguments

Terraform CLI defines the following optional arguments for variable declarations:

- [`default`][inpage-default] - A default value which then makes the variable optional.
- [`type`][inpage-type] - This argument specifies what value types are accepted for the variable.
- [`description`][inpage-description] - This specifies the input variable's documentation.
- [`validation`][inpage-validation] - A block to define validation rules, usually in addition to type constraints.

> - [`sensitive`][inpage-sensitive] - A boolean value that can limit Terraform UI output when the variable is used in configuration, to avoid accidental disclosure in logs or other UI.

Reviewer: Nitpick: `sensitive` is  an argument, not a value. I might say "An argument that can limit ..."

Author: The "value" here refers to the right side of the argument. I'm worried that if I change it here, it doesn't make sense for ex. `default`

should be at least one full sentence explaining the constraint that failed,
using a sentence structure similar to the above examples.

### Sensitive

[inpage-sensitive]: #sensitive

-> This feature was introduced in Terraform CLI v0.14.0.

> The `sensitive` argument on a variable block is a boolean value that, when provided, limits the output of the Terraform `plan` or `apply` when that variable is used. A provider can define [an attribute as sensitive](/docs/extend/best-practices/sensitive-state.html#using-the-sensitive-flag), which prevents the value of that attribute from being displayed in logs or regular output. The `sensitive` argument on variables allows users to replicate this behavior for values in their configuration, by defining a variable as `sensitive`.

Reviewer: I might rephrase this sentence a bit, inspired by our sensitive outputs documentation (https://www.terraform.io/docs/configuration/outputs.html):

Setting a variable as `sensitive` argument prevents Terraform from showing its value in the `plan` and `apply` output.

Once you have defined a sensitive variable, using it throughout your configuration will obfuscate the value from display in output:

```
variable "user_information" {
  type = object({
    name    = string
    address = string
  })
  sensitive = true
}

resource "some_resource" "a" {
  name    = var.user_information.name
  address = var.user_information.address
}
```

Running `terraform apply`:

```
Terraform will perform the following actions:

  # some_resource.a will be created
  + resource "some_resource" "a" {
      + name    = (sensitive)
      + address = (sensitive)
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

#### Cases where Terraform may disclose a sensitive variable

> Variable values marked as sensitive will display in state. Much like provider-side sensitive values, designating a value as sensitive only limits its display in logs or output, not in state.

Reviewer: Is this saying they are stored in state (which I know they are) or that they will be displayed in `state show`, or both? "display in state" might seem ambiguous to some readers.

Author: Both

> Similarly, `sensitive` argument does not have an impact in other Terraform commands such as `console` or `show`. Those commands are meant to ex as this argument is intended to reduce exposure of data in, for example, external logs or aggregation.

Reviewer: > Those commands are meant to ex as this 
Not quite sure what is intended here?

Author: Oops, just a mistake. Thanks for flagging this!

> Defining sensitivity at the variable level is a configuration-centered concept, and values are sent to providers without any obfuscation; as such, a provider error could disclose a value depending on if the value is included in the error message. For example, if a provider returns an error with `"Invalid value 'foo' for field"`, the value will be displayed in output as a consequence of this error.

Reviewer: ```suggestion
A ` sensitive` variable level is a configuration-centered concept, and values are sent to providers without any obfuscation. A provider error could disclose a value if that value is included in the error message. For example, a provider might return the following error even if "foo" is a sensitive value: `"Invalid value 'foo' for field"`
```

This paragraph could use a bit more of a direct and active tone though my suggestion could also use some work :)

If a resource attribute is used as, or part of, the provider-defined resource id, an `apply` will disclose the value. In the example below, the `prefix` attribute has been set to a sensitive variable, but then that value ("jae") is later disclosed as part of the resource id:

```
  # random_pet.animal will be created
  + resource "random_pet" "animal" {
      + id        = (known after apply)
      + length    = 2
      + prefix    = (sensitive)
      + separator = "-"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

...

random_pet.animal: Creating...

> random_pet.animal: Creation complete after 0s [id=jae-known-mongoose]

Reviewer: This is a great example!
